app/repositories/SessionRepository.scala

```diff
@@ -57,43 +59,49 @@ class SessionRepository @Inject()(formId: String, mongo: ReactiveMongoComponent)
     saveOrUpdate[A](data)
 
   override def saveOrUpdate[A](data: A)(implicit wts: Writes[A], hc: HeaderCarrier): Future[Unit] =
-    for {
-      sessionId <- getSessionId
-      _ <- collection.update(
-            BSONDocument("_id" -> BSONString(sessionId)),
-            BSONDocument(
-              "$set"         -> BSONDocument(s"data.$formId" -> Json.toJson(data)),
-              "$setOnInsert" -> BSONDocument("createdAt"     -> BSONDateTime(System.currentTimeMillis))
-            ),
-            upsert = true
-          )
-    } yield {
-      ()
+    Mdc.preservingMdc {
+      for {
+        sessionId <- getSessionId
+        _ <- collection.update(
+          BSONDocument("_id" -> BSONString(sessionId)),
+          BSONDocument(
+            "$set" -> BSONDocument(s"data.$formId" -> Json.toJson(data)),
+            "$setOnInsert" -> BSONDocument("createdAt" -> BSONDateTime(System.currentTimeMillis))
+          ),
+          upsert = true
+        )
+      } yield {
+        ()
+      }
     }
 
   override def get[A](implicit rds: Reads[A], hc: HeaderCarrier): Future[Option[A]] =
-    for {
-      sessionId   <- getSessionId
-      maybeOption <- findById(sessionId)
-    } yield {
-      maybeOption
-        .map(_.data \ formId)
-        .flatMap(x =>
-          x match {
-            case JsDefined(value) => Some(value.as[A])
-            case JsUndefined()    => None
-        })
+    Mdc.preservingMdc {
+      for {
+        sessionId <- getSessionId
+        maybeOption <- findById(sessionId)
+      } yield {
+        maybeOption
+          .map(_.data \ formId)
+          .flatMap(x =>
+            x match {
+              case JsDefined(value) => Some(value.as[A])
+              case JsUndefined() => None
+            })
```

Comment:
I know you didn't write this originally, but it struck me as particularly ugly piece of code

```suggestion
          maybeOption.flatMap(sd => (sd.data \ formId).asOpt[A])
```